Fix lossless formats in PortableCompressedTexture2D

nklbdev opened the pull request:
This is a right copy of my [previous wrong pull-request](https://github.com/godotengine/godot/pull/77047)

Sorry for my bad English(((

`PortableCompressedTexture2D` have a problem with creating from PNG images and some other little problems.
I want to use this resource type as the way to import animations from Aseprite in my [Aseprite importers plugin bundle](https://github.com/nklbdev/godot-4-aseprite-importers)

For now I am creating one PNG image near every `*.aseprite`-file that imported. Then call some methods to update filesystem and import the image. Then plugin is continuing it's work and embedding new image in an imported resource.

`PortableCompressedTexture2D` is a beautiful way to embed PNG data silently in `SpriteFrames` or other resources. But it is not usable now. Method `create_from_image` cannot load PNG image and shows errors in console:

```
WARNING: Not a PNG file
     at: PNGDriverCommon::check_error (drivers\png\png_driver_common.cpp:56)
ERROR: Condition "!success" is true. Returning: ERR_FILE_CORRUPT
   at: PNGDriverCommon::png_to_image (drivers\png\png_driver_common.cpp:69)
ERROR: Condition "err" is true. Returning: Ref<Image>()
   at: ImageLoaderPNG::load_mem_png (drivers\png\image_loader_png.cpp:64)
ERROR: It's not a reference to a valid Image object.
   at: (.\core/io/image.h:426)
ERROR: Condition "err" is true. Returning: Ref<Image>()
   at: _jpegd_mem_loader_func (modules\jpg\image_loader_jpegd.cpp:131)
ERROR: It's not a reference to a valid Image object.
   at: (.\core/io/image.h:426)
ERROR: Condition "p_width <= 0 || p_width > 16384" is true.
   at: GLES3::TextureStorage::texture_set_size_override (drivers\gles3\storage\texture_storage.cpp:1053)
```

This is because PNG image have not only PNG prefix `‰PNG`, but also [Godot's own `PNG ` prefix](https://github.com/godotengine/godot/blob/ffd32a244b43ff58c13819c2debf8cf3b58ecbdc/drivers/png/image_loader_png.cpp#L80)


I understand that this solution is not ideal.
Theoretically, there could be such a sequence of bytes PNG  in this memory location for some other reason. Not because Godot added it there. I don't know how to be completely sure of this. If you know such a criteria, please tell me and I will correct the code.

Result: It works! Now I use `PortableCompressedTexture2D` and save it directly in imported resource - `SpriteFrames` or `PackedScene`!

https://github.com/godotengine/godot/assets/7024016/79853566-1df0-4ac5-962b-49667a034b3e

nklbdev commented on scene/resources/texture.cpp:
```diff
@@ -370,6 +371,15 @@ void PortableCompressedTexture2D::_set_data(const Vector<uint8_t> &p_data) {
 				data += 4;
 				data_size -= 4;
+
+				// Skip possible Godot png tag "PNG " before pure PNG data
```
[Godot own PNG prefix](https://github.com/godotengine/godot/blob/ffd32a244b43ff58c13819c2debf8cf3b58ecbdc/drivers/png/image_loader_png.cpp#L80) must been skipped

AThousandShips replied:
This comment should also be correctly formatted

nklbdev replied:
Done)

nklbdev commented on scene/resources/texture.cpp:
```diff
@@ -406,6 +416,7 @@ void PortableCompressedTexture2D::_set_data(const Vector<uint8_t> &p_data) {
 	}
 
 	image_stored = true;
+	size_override = size;
```
Right handling of `size_override`. Because this is produce an error if it uninitialized (`Vector2i.Zero`)

AThousandShips commented on scene/resources/texture.cpp:
```diff
@@ -438,12 +449,17 @@ void PortableCompressedTexture2D::create_from_image(const Ref<Image> &p_image, C
 		case COMPRESSION_MODE_LOSSY: {
+			bool lossless_force_png = GLOBAL_GET("rendering/textures/lossless_compression/force_png") ||
+					!Image::_webp_mem_loader_func; // WebP module disabled.
+			bool use_webp = !lossless_force_png && p_image->get_width() <= 16383 && p_image->get_height() <= 16383; // WebP has a size limit
```
```suggestion
			bool use_webp = !lossless_force_png && p_image->get_width() <= 16383 && p_image->get_height() <= 16383; // WebP has a size limit.
```
Comment format nitpick

nklbdev replied:
Ok) this is just a copy from [`resource_importer_texture.cpp`](https://github.com/godotengine/godot/blob/95c8216923fc9353ac50f469836ae149cb196e83/editor/import/resource_importer_texture.cpp#L247)

nklbdev commented:
I'll need add some changes to save `PortableCompressedTexture2D` to `*.pctex`-file

nklbdev commented:
I've added an image file format detection to avoid trying to load unsuitable formats, red console messages, and exceptions.
This code has been thoroughly tested and conforms to the specifications of these file formats.

nklbdev commented:
I found an error in [`servers/rendering_server.cpp` on line 442](https://github.com/nklbdev/godot/blob/803256d5c9b0f804a04ef2e2fc48093c0aa1e204/servers/rendering_server.cpp#LL442C9-L442C9) and many other lines:
An integer value in the range 0 to 65535 is casted to `int16_t`, which will accommodate the range from -32767 to 32767. This causes the application to crash.

Now I'll try to fix it.

nklbdev commented on core/io/image.cpp:
```diff
@@ -3929,22 +3929,78 @@ void Image::renormalize_rgbe9995(uint32_t *p_rgb) {
+			}
+			break;
+		default: // FILE_FORMAT_UNKNOWN
+			initialize_data(1, 1, false, FORMAT_RGBA8);
```
During the checks, an interesting phenomenon was discovered: all the checks pass except for one.

It seems that in the assembly for Linux, an icon in an invalid format is loaded into the DisplayServerX11. Regardless, the image is created. I can't find out what the file is and I can't check what image was created from it. But it certainly turns out to be non-empty.

Adding my format checks caused this image to stop being generated (had a width and a height, but didn't have a single byte in the data block). This caused the application to crash in this test.

That's why I put a 1 pixel transparent image initialization here to avoid such errors.

Here is a listing of the log so that you can find the damaged image:

### 🐧 Linux / Editor with doubles and GCC sanitizers (target=editor, tests=yes, dev_build=yes, precision=double, use_asan=yes, use_ubsan=yes, linker=gold)

```
ERROR: FATAL: Index p_index = 2 is out of bounds (((Vector<T> *)(this))->_cowdata.size() = 2).
   at: operator[] (./core/templates/vector.h:52)

================================================================
handle_crash: Program crashed with signal 4
Engine version: Godot Engine v4.1.dev.custom_build (db8d3213c200ac922dd7ad0199127b13d674d006)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(+0x3cb0bac4) [0x55c042a3eac4] (??:?)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7efed6d93090] (??:0)
[3] VectorWriteProxy<long>::operator[](int) (??:?)
[4] DisplayServerX11::set_icon(Ref<Image> const&) (??:?)
[5] Main::start() (??:?)
[6] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(main+0x3da) [0x55c042a3e1c3] (??:?)
[7] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7efed6d74083] (??:0)
[8] ./bin/godot.linuxbsd.editor.dev.double.x86_64.san(_start+0x2e) [0x55c042a3dd2e] (??:?)
-- END OF BACKTRACE --
================================================================
Aborted (core dumped)
FATAL ERROR: Godot has been crashed.
Error: Process completed with exit code 52.
```

nklbdev commented on core/io/image.cpp:
```diff
@@ -3929,22 +3929,79 @@ void Image::renormalize_rgbe9995(uint32_t *p_rgb) {
 }
 
+Image::FileFormat Image::detect_file_format(const uint8_t *p_data, int p_len) {
+	size_t len = p_len < 0 ? 0 : unsigned(p_len);
```
Sometimes the image creation function is called without specifying the length of the data. In this case, the data length is set to -1. Therefore, I made length comparisons only if a positive data length is explicitly specified.

nklbdev commented:
Done

nklbdev commented:
In a week or two I will finish developing the same set of importers for animations from Krita.  They also use PortableCompressedTexture2D.

clayjohn commented on servers/rendering_server.cpp:
```diff
@@ -402,9 +402,9 @@ Error RenderingServer::_surface_set_data(Array p_arrays, uint32_t p_format, uint
-					int16_t vector[2] = {
-						(int16_t)CLAMP(res.x * 65535, 0, 65535),
-						(int16_t)CLAMP(res.y * 65535, 0, 65535),
+					uint16_t vector[2] = {
+						(uint16_t)CLAMP(res.x * 65535, 0, 65535),
+						(uint16_t)CLAMP(res.y * 65535, 0, 65535),
```
The changes to this file should not be included in this PR as they are unrelated. 

Could please remove them from this PR, and open a new PR with only these changes (along with a link to the bug or description of the bug if it hasn't been reported yet).

nklbdev replied:
I'd like to push these changes to another Pull Request, but without them, the current pull request has an error in the automatic check [`Linux / Editor with doubles and GCC sanitizers`](https://github.com/godotengine/godot/actions/runs/4970081774/jobs/8893729737?pr=77047). I mentioned it in a previous copy of this pull request.

It appears because the engine is trying to use an image that did not match one of the possible types. In this case, the old code makes 3 attempts to initialize the image. One of these functions, despite the internal error, still initializes a small empty image. This masks the problem.

When using my image type detector, none of these functions are called and the image remains uninitialized. Then the value -1 is passed as the data length through the `int` type, which is cast to an unsigned type and becomes a very large number. This generates an error that fails the check I mentioned above.


https://github.com/godotengine/godot/pull/77047#issuecomment-1546793960

clayjohn replied:
I don't know how that can be as this PR only changes image/texture functionality and these lines of code are only run when creating a mesh surface. They have nothing to do with images/textures

nklbdev replied:
Without these lines of code, the check that I described above will not pass. Correct reading of the image data leads to the discovering of a hidden problem.

As a result, the code works as it should - the wrong image is really not loaded. The image object remains uninitialized (that's right).

As a result of this, here, in the `RenderingServer`, the `x` and `y` components get value 1. This results in a large number (65535) in the expression (it is OK), and a negative number after the wrong cast (it is not OK). This causes Godot to fail on this check.

Therefore, I cannot remove these changes from my PR. And we need to find and replace the image that Godot is trying to load.

clayjohn commented on core/io/image.cpp:
```diff
@@ -3929,22 +3929,79 @@ void Image::renormalize_rgbe9995(uint32_t *p_rgb) {
+Image::FileFormat Image::detect_file_format(const uint8_t *p_data, int p_len) {
+	size_t len = p_len < 0 ? 0 : unsigned(p_len);
+
+	static const uint8_t png_file_header[] = { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A }; // \211PNG\r\n\032\n
+	static const size_t png_file_header_size = sizeof(png_file_header);
+	if ((!len || len > png_file_header_size) && memcmp(p_data, png_file_header, png_file_header_size) == 0) {
+		return FILE_FORMAT_PNG;
+	}
```
I have a feeling this code should be in the respective modules for the image type. The current design of Image avoids having filetype-specific code by offloading all the work to the various modules. This is in line with our engine best practice to ["prefer local solutions"](https://docs.godotengine.org/en/latest/contributing/development/best_practices_for_engine_contributors.html#prefer-local-solutions). 

We are especially strict with core types (like Image), because if they become bloated and messy, then the entire engine becomes bloated and messy.

nklbdev replied:
This method is used in the image constructor, which takes only raw data as input.

This constructor is only used in one place - [texture.cpp PortableCompressedTexture2D::_set_data](https://github.com/godotengine/godot/blob/a83eb16fba5bb3da086b41cbd79e6f95b09eb8ee/scene/resources/texture.cpp#LL374C31-L374C31). I'd really like to move the image type detection function to a different location. Can you please tell me what is the best place for this?

clayjohn replied:
I would add it to the respective loader functions i.e. ``_png_mem_loader_func``. If the format is the wrong type those functions can exit early which will leave the image empty so that the next format can be checked.

nklbdev replied:
done

clayjohn commented on scene/resources/texture.cpp:
```diff
@@ -370,6 +371,15 @@ void PortableCompressedTexture2D::_set_data(const Vector<uint8_t> &p_data) {
 				data_size -= 4;
+
+				// Skip possible Godot png tag "PNG " before pure PNG data.
+				static const uint8_t godot_png_prefix[] = { 0x50, 0x4E, 0x47, 0x20 };
+				static const size_t godot_png_prefix_size = sizeof(godot_png_prefix);
+				if (memcmp(data, godot_png_prefix, godot_png_prefix_size) == 0) {
+					data += godot_png_prefix_size;
+					data_size -= godot_png_prefix_size;
+				}
```
How well does this interact with WebP files? I don't see any type checking to ensure that we aren't breaking WebP files in the process.

nklbdev replied:
For WebP format, this code is safe: WebP image data starts with `RIFF` characters

I also don't like this kind of interference with the raw data at this point, where it's not yet clear what format this image actually has. I would prefer to know the format first, and then perform such manipulations.

If you know where to move this code, please tell me! I'll be very happy.

clayjohn replied:
Its probably a good idea to keep the implementation similar to CompressedTexture2D in that case. CompressedTexture2D stores a ``DataFormat`` enum which is used (among other things) to signal what format the texture is using. That way we would know whether PNG or WEBP is used and can load the information accordingly.

nklbdev replied:
I can't do this because `CompressedTexture2D` determines the format by the first byte in the [`load_image_from_file`](https://github.com/godotengine/godot/blob/46424488edc341b65467ee7fd3ac423e4d49ad34/scene/resources/texture.cpp#L657C21-L657C37) method, where file is the `ctex` file created by the `ResourceImporterTexture` class in method [`save_to_ctex_format`](https://github.com/godotengine/godot/blob/46424488edc341b65467ee7fd3ac423e4d49ad34/editor/import/resource_importer_texture.cpp#L249C23-L249C23).

A `PortableCompressedTexture2D` is created not from a file, but from an `Image` instance, which does not contain this information. And this information cannot be added to it, as this will damage so many other places in the code.

If you know how to do this, I will be very glad to your advice!

clayjohn commented on scene/resources/texture.cpp:
```diff
@@ -438,12 +449,17 @@ void PortableCompressedTexture2D::create_from_image(const Ref<Image> &p_image, C
 	switch (p_compression_mode) {
 		case COMPRESSION_MODE_LOSSLESS:
 		case COMPRESSION_MODE_LOSSY: {
+			bool lossless_force_png = GLOBAL_GET("rendering/textures/lossless_compression/force_png") ||
```
Is this necessary to fix anything? I wasn't involved in the decision to switch to WebP, but I know that the contributors who did agonized over the decision and concluded that it was better to go all in on WebP. Accordingly, I would stick to WebP if at all possible.

nklbdev replied:
This fragment of code is copied as is from [resource_importer_texture.cpp](https://github.com/godotengine/godot/blob/95c8216923fc9353ac50f469836ae149cb196e83/editor/import/resource_importer_texture.cpp#L247).

Is it worth moving this common fragment into a separate function so that both of these classes can use it?

clayjohn replied:
Ah, in that case, it is probably fine as is.

clayjohn reviewed (commented):
I left a few comments with my thoughts and some suggestions. Please let me know if you have a question about any of my suggestions.

nklbdev commented:
Thanks @clayjohn, I completely rework the changes as per your advice and it worked out pretty well.

The `PortableCompressedTexture2D`, when writing data, specifies the type of data in the first 4 bytes (`int32`). When data is read, these bytes are read to `CompressedTexture2D.DataFormat` and the load procedure can choose image loading function.

Also, I made a separate function that can "unpack" PNG not from the buffer, but from the memory fragment by the pointer.

bitsawer commented:
I was fixing some other issues with PortableCompressedTexture2D and while making my PR (https://github.com/godotengine/godot/pull/79579) I noticed this one. Looks good, but doesn't adding a new flag to the header break compatibility? PortableCompressedTexture2D is a resource, so everyone who has saved them in previous Godot versions will now get invalid resources as the code doesn't have any versioning.

nklbdev commented:
@bitsawer, yes, it will break compatibility. All resources containing PortableCompressedTexture2D must be reimported(((

nklbdev commented:
@bitsawer, please tell me how to avoid this if you know.

caimantilla commented:
This is pretty important to fix IMO. The current implementation is broken. Even if it breaks compatibility, it was already a problem to begin with... Other things that work fine as-is will have compatibility-breaking changes in the future, I think it's justified to do the same with PortableCompressedTexture2D which _doesn't_ work fine.

nklbdev commented:
I am sure that this task can be done better than I did. But I will need your help, because I write mainly in CSharp, and have not touched CPlusPlus for a very long time.
Perhaps there is a way to avoid breaking backward compatibility.

Also, I think it is possible to give import plugins the ability to import images into CompressedTexture (.ctex) format. And the ability to embed this resource in other resources. In this case, the need for a PortableCompressedTexture will disappear completely.

caimantilla commented:
> I am sure that this task can be done better than I did. But I need your help, because I write mainly in C#, and have not touched C++ for a very long time.
> 
> Perhaps there is a way to avoid breaking backward compatibility.
> 
> Also, I think it is possible to give import plugins the ability to import images into CompressedTexture (.ctex) format. And the ability to embed this resource in other resources. In this case, the need for a PortableCompressedTexture will disappear completely.

The reason why compatibility might even matter is that its intended usage is for user-generated content which can't just be reimported. That it's useful for plugins is nice but it still has a purpose

nklbdev commented:
> The reason why compatibility might even matter is that its intended usage is for user-generated content which can't just be reimported. That it's useful for plugins is nice but it still has a purpose

Now I have managed to generate a compilation database. It helps a lot to navigate through the code.

I rebase my branch and fix the conflicts. Now I'm debugging the code.

I'll try to come up with a way to keep backwards compatibility.

nklbdev committed:
Fix lossless formats in PortableCompressedTexture2D

Update scene/resources/portable_compressed_texture.cpp

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

nklbdev commented:
Previously, I tried to add 4 bytes in front of the data in the file, which encoded the data type (IMAGE, PNG, WEBP and BASIS_UNIVERSAL). Now I have made it so that the data type is encoded in the existing 3rd and 4th bytes.

Also I've added a default value of 0 for the undefined data type for backwards compatibility. In this case, the data reading will proceed according to the previous scenario.

At the same time, the compression mode is now encoded not in the first four, but in the first two bytes. This is ok: this value fits perfectly even in 1 byte.

I also added the ability to preview PortableCompressedTexture2D to the EditorInspectorPluginTexture.

nklbdev commented:
Please tell me, is it OK?

Bytes before (4 bytes for compression mode):
1. small bits of compression mode
2. small-mid bits of compression mode
3. mid-big bits of compression mode
4. big bits of compression mode

Bytes after (2 bytes for compression mode and 2 bytes for data type):
1. small bits of compression mode
2. big bits of compression mode
3. small bits of data type
4. big bits of data type

fire commented:
Commentary:

1. Concerns that this needs to keep backwards compatibility
2. Feature freeze is upcoming.
3. Improves the situation

AThousandShips commented on scene/resources/compressed_texture.h:
```diff
@@ -40,10 +40,11 @@ class CompressedTexture2D : public Texture2D {
-		DATA_FORMAT_PNG,
-		DATA_FORMAT_WEBP,
-		DATA_FORMAT_BASIS_UNIVERSAL,
+		DATA_FORMAT_UNDEFINED = 0,
```
On the other hand it can cause confusion, and clutters

nklbdev replied:
Fine. Now I will remove these values and rebase my branch to the current master. But if the order of these names is changed in the future, this will break compatibility with already created resources.

fire commented:
Hiya, sorry this got lost. I'll see what I can to nudge this.

nklbdev commented:
> Hiya, sorry this got lost. I'll see what I can do to nudge this.

Thank you!
If anyone has any ideas on how to implement this better, I will be very happy.

lyuma commented on scene/resources/portable_compressed_texture.cpp:
```diff
@@ -122,7 +138,7 @@ void PortableCompressedTexture2D::create_from_image(const Ref<Image> &p_image, C
 
 	buffer.resize(20);
-	encode_uint32(p_compression_mode, buffer.ptrw());
+	encode_uint16(p_compression_mode, buffer.ptrw());
```
Contrary to reasonable security practices, Vector::resize() is *not zeroed* in Godot C++. This is done to keep reviewers and contributors on their toes, and I certainly am standing on my toes as I write this.

While there are several cases below that write to these extra bytes, I'm having trouble following the code flow below. I would rather guarantee that something is written here by default, even if it is later overwritten in all reasonable cases.
```suggestion
	encode_uint16(p_compression_mode, buffer.ptrw());
	encode_uint16(CompressedTexture2D::DATA_FORMAT_UNDEFINED, buffer.ptrw() + 2);
```

lyuma replied:
Looks good now.

lyuma commented on scene/resources/portable_compressed_texture.cpp:
```diff
@@ -99,6 +114,7 @@ void PortableCompressedTexture2D::_set_data(const Vector<uint8_t> &p_data) {
 	}
 
 	image_stored = true;
+	size_override = size;
```
Good catch. I'm surprised it worked without this before: size_override would have been initialized to zero.

lyuma reviewed (approved):
I think it looks great. The changes made in September address the original concern about backwards compatibility breakage ("2 bytes for compression mode and 2 bytes for data type") so I see no problems with it. Just one nitpick.

Sorry we didn't get to this before the 4.2 feature freeze, but now that the 4.3 merge window is open, I see no problem with merging this.

nklbdev commented:
@lyuma, wonderful! Now I will rebase and make your edits. I should warn you that I am not a C++ professional and usually work with C#. Therefore, be attentive to the code that I made. I may not know some subtleties.

lyuma commented:
I also removed breaks compat label since this does not break backwards compatibility. I don't think any of the concerns raised prior to september are still valid. Do you think you can double check the past review comments and click Resolve Conversation on the discussions which are no longer applicable?

nklbdev commented:
@lyuma, I checked and resolved all the discussions that I could resolve.

Keep in mind:
- Textures created by a previous version of the engine will be able to be opened by this version.
- But textures created by this version of the engine will no longer be able to open normally with the previous version.

akien-mga commented on scene/resources/portable_compressed_texture.cpp:
```diff
@@ -30,6 +30,8 @@
 
 #include "portable_compressed_texture.h"
 
+#include "compressed_texture.h"
+#include "core/config/project_settings.h"
 #include "core/io/marshalls.h"
 #include "scene/resources/bit_map.h"
```
We use absolute paths even when in the same folder (aside from the .h file matching the .cpp).
```suggestion
#include "core/config/project_settings.h"
#include "core/io/marshalls.h"
#include "scene/resources/bit_map.h"
#include "scene/resources/compressed_texture.h"
```

akien-mga commented:
Looks good! Could you squash the commits? See [PR workflow](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase) for instructions.

nklbdev commented:
I was a little mistaken. Wait a little, now I'll fix everything.

nklbdev commented:
Yes, I just rushed a little and GitExtensions executed a command that I didn't want. Now everything is fine: the branch is inherited from the most recent commit in the master branch

akien-mga commented:
Thanks!

nklbdev commented:
Thanks, too!  Happy New Year!

akien-mga commented on scene/resources/compressed_texture.h:
```diff
@@ -40,6 +40,7 @@ class CompressedTexture2D : public Texture2D {
 
 public:
 	enum DataFormat {
+		DATA_FORMAT_UNDEFINED,
```
This apparently broke compatibility for existing compressed textures, as it's reordering all enum values. New enum values should go to the end of the enum to avoid breaking compatibility.

(Haven't tested yet to confirm but I got multiple reports of a compat breakage on compressed textures introduced in today's merges.)

nklbdev replied:
I'll check it now

nklbdev replied:
@akien-mga, how can I push a fix?

Zireael07 replied:
IIRC the best way to go about it (since this PR has been merged) is to make any fixes to a separate branch and PR them separately

nklbdev replied:
Okay, I'll make a new pull request now.

The problem is that the `UNDEFINED` image format is the format that all the old images are in. That's why I can't put it at the end of the list.

nklbdev replied:
I created a new pull-request: https://github.com/godotengine/godot/pull/86835

nklbdev replied:
I'm very sorry that this happened. I want to do everything to minimize the consequences. Perhaps even rollback this pull request if necessary
